test/support/utils_sycl.h

```diff
@@ -140,28 +140,90 @@ make_new_policy(_Policy&& __policy)
 // create the queue with custom asynchronous exceptions handler
 static auto my_queue = sycl::queue(default_selector, async_handler);
 
-// Invoke op(policy,rest...) for each possible policy.
+// check fp16/fp64 support by a device
+template<typename T>
+bool has_type_support(const sycl::device&) { return true; }
+
+template<>
+inline bool has_type_support<double>(const sycl::device& device)
+{
+    return device.has(sycl::aspect::fp64);
+}
+
+template<>
+inline bool has_type_support<sycl::half>(const sycl::device& device)
+{
+    return device.has(sycl::aspect::fp16);
+}
+
+template <typename, typename = void>
+struct has_value_type : ::std::false_type {};
+
+template <typename T>
+struct has_value_type<T, ::std::void_t<typename std::iterator_traits<T>::value_type>> : ::std::true_type {};
+
+template<typename T>
+std::enable_if_t<!has_value_type<T>::value, bool> has_iter_type_support(const sycl::device& device)
```

Comment:
Seems like we don't support zip_iterator and furthermore, other types that are semantically wrappers (like tuples, pairs, opitonal, etc.) Need to think what we do with that.

Reply:
Seems like the simplest way is to check if it's `zip_iterator` and write the specialization for that taking the value_type (that is unspecified) and then check each type of tuple. The rest can be omitted if we know for sure that it doesn't break something

Reply:
Another option is to return to the initial approach, where we explicitly specify types to check on a test level. We should modify majority of the tests in that case.

We can also highlight that issue for DPC++ compiler or SYCL spec and ask if we can check a custom type directly instead of using sycl::aspect::fp16/fp64 traits.

Reply:
Meanwhile, I like the proposed "simplest" way.

Reply:
I left TODO for now, because we do not test zip_iterator with double/half types on a device.